### ISSUE TITLE
unordered version of temporal community detection

### ISSUE THREAD

**Issue by mb3152** [vtraag/leidenalg#67]
Would it be easy / possible to build a wrapper that links nodes across all layers? This would be good for say, a bunch of different brain networks from different subjects. So, no temporal order; you want a node to be linked to itself across all layers, not just the graph before and after it in the array of graphs. They support this in the GenLouvain MATLAB library.

**Comment by vtraag** [comment]
You can do this using the function [`slices_to_layers`](https://leidenalg.readthedocs.io/en/stable/reference.html#leidenalg.slices_to_layers). This is just a generic function where a coupling graph allows you to specify how nodes across different slices should be connected. Each slice is a node of the coupling graph, and the connections between the nodes of the coupling graph indicate how the slices should be connected.

For example, for connecting three slices all between each other, you can use

```python
H_coupling = ig.Graph.Full(3)
H_coupling.vs['slice'] = [H1, H2, H3]
H_coupling.es['weight'] = 1
G_layers, G_interslice, G_full = leidenalg.slices_to_layers(H_coupling)
```

You can then create the necessary partition and optimise them, this is also explained in the more detailed section on converting [slices to layers](https://leidenalg.readthedocs.io/en/stable/multiplex.html#slices-to-layers).

Is this a situation that many people will encounter? If so, I could add a convenience function, similar to [`find_partition_temporal`](https://leidenalg.readthedocs.io/en/stable/reference.html#leidenalg.find_partition_temporal)

**Comment by mb3152** [comment]
Ah, I see how that works. Just to ensure I am doing it correctly, H1, H2, and H3 will only have links between layers from a node to itself, not to other nodes?

I assume many people would use it for any unordered multi-slice network. Quite common in human neuroscience. Would greatly appreciated a convenience function!

Thank you for your time.

**Comment by vtraag** [comment]
>  H1, H2, and H3 will only have links between layers from a node to itself, not to other nodes?

Yep, that's right, it only connects nodes with similar node identifiers. The attribute that is used to identify "the same node" across multiple slices can be specified using the `vertex_id_attr`. See the mentioned docs more explanation.

> I assume many people would use it for any unordered multi-slice network. Quite common in human neuroscience. Would greatly appreciated a convenience function!

I'll think about an option to make this type of thing simpler! I'll leave issue open to remind me.

**Comment by mb3152** [comment]
thank you!!

**Comment by mb3152** [comment]
Sorry, I am not sure I am doing this correctly. I am expecting a 25(graphs) x 400(nodes) membership array.

```python
	H_coupling = ig.Graph.Full(25)
	H_coupling.vs['slice'] = graphs
	H_coupling.es['weight'] = .5
	layers, interslice_layer, G_full= la.slices_to_layers(H_coupling)
	partitions = [la.CPMVertexPartition(H, node_sizes='node_size',weights='weight',resolution_parameter=1)for H in layers]
	interslice_partition = la.CPMVertexPartition(interslice_layer, resolution_parameter=0,node_sizes='node_size', weights='weight')
	diff = optimiser.optimise_partition_multiplex(partitions + [interslice_partition])
```

**Comment by mb3152** [comment]
I tried to edit find_partition_temporal by passing it my (layers, interslice_layer, G_full) from above instead of making it in the function. I get a key error at the "membership_slice = [membership[(slice_idx, v[vertex_id_attr])] for v in H.vs]" line though.

**Comment by mb3152** [comment]
Can you explain the usage of CPMVertexPartition with a resolution parameter of zero? Does this partition solution matter to the final partition?

**Comment by vtraag** [comment]
> Can you explain the usage of CPMVertexPartition with a resolution parameter of zero? Does this partition solution matter to the final partition?

This corresponds to the interslice coupling. There is no cost associated (i.e. no penalty for bringing nodes together in the same community), hence the resolution parameter should be zero. In principle one could also use a `RBERVertexPartition` or a `RBConfigurationVertexPartition`, as long as the resolution parameter is zero. This corresponds to the d_ij C_jsr in Eq. (3) in [Mucha et al. (2020)](http://dx.doi.org/10.1126/science.1184819) 

> Also, I am running this on a 400 node network, density is 1 percent, with 900 layers. Any idea how long this should take? I submitted Friday, still running this weekend. I assume that's a bit longer than expected. Am I doing it wrong?

Note that the networks grow quite big very fast. With 900 slices of 400 nodes, you are running the algorithm on 900 layers of 900*400 = 360 000 nodes. Nonetheless, the current implementation is less efficient for many layers. There have been some attempts at making this faster for the earlier Louvain implementation, see https://github.com/vtraag/louvain-igraph/pull/34, but this was only implemented for modularity and did not generalize easily over all supported quality functions with substantial effort. Nonetheless, you may be interested in checking this out. You could ask @wweir827 or @ragibson if they perhaps have an updated version of their improvement.
